File: leetcode/dp/2313_Minimum_Flips_in_Binary_Tree_to_Get_Result/codes/dp.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def minimumFlips(self, root: Optional[TreeNode], result: bool) -> int:
        def dfs(node):
            if not node:
                return {True: float("inf"), False: float("inf")}

            if not node.left and not node.right: 
                val = bool(node.val)
                return {val: 0, not val: 1}

            if node.val == 5: 
                child = node.left or node.right
                child_dp = dfs(child)
                return {
                    True: child_dp[False],
                    False: child_dp[True]
                }

            left_dp = dfs(node.left)
            right_dp = dfs(node.right)

            dp = {True: float("inf"), False: float("inf")}
            for l_val in [False, True]:
                for r_val in [False, True]:
                    if node.val == 2:  
                        res = l_val or r_val
                    elif node.val == 3:  
                        res = l_val and r_val
                    elif node.val == 4:  
                        res = l_val ^ r_val
                    else:
                        continue  

                    cost = left_dp[l_val] + right_dp[r_val]
                    dp[res] = min(dp[res], cost)

            return dp

        dp = dfs(root)
        return dp[result]

File: leetcode/dp/2313_Minimum_Flips_in_Binary_Tree_to_Get_Result/codes/dp.cpp
#include <vector>
#include <array>
#include <algorithm>
#include <climits>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    array<int, 2> dfs(TreeNode* node) {
        if (!node) return {INT_MAX, INT_MAX};

        if (!node -> left && !node -> right) {
            int val = node -> val;
            return val == 0 ? array<int, 2>{0, 1} : array<int, 2>{1, 0};  
        }

        if (node -> val == 5) { 
            TreeNode* child = node -> left ? node -> left : node -> right;
            auto c = dfs(child);
            return {c[1], c[0]};
        }

        auto l = dfs(node -> left);
        auto r = dfs(node -> right);
        array<int, 2> dp = {INT_MAX, INT_MAX};

        for (int lv = 0; lv <= 1; lv++) {
            for (int rv = 0; rv <= 1; rv++) {
                int res = 0;
                if (node -> val == 2) res = lv | rv;
                else if (node -> val == 3) res = lv & rv;
                else if (node -> val == 4) res = lv ^ rv;

                dp[res] = min(dp[res], l[lv] + r[rv]);
            }
        }

        return dp;
    }

    int minimumFlips(TreeNode* root, bool result) {
        auto dp = dfs(root);
        return result ? dp[1] : dp[0];
    }
};
